components/events/tests/event_pipeline.test.cpp

```diff
@@ -0,0 +1,313 @@
+	TestEventPipeline pipeline{};
+
+	REQUIRE(pipeline.once_count() == 0);
+
+	pipeline.add_once(std::make_unique<TypedEventPipelineStage<TestEvent>>());
+
+	REQUIRE(pipeline.once_count() == 1);
+}
+
+TEST_CASE("register then stage", "[events]")
+{
+	TestEventPipeline pipeline{};
+
+	REQUIRE(pipeline.always_count() == 0);
+
+	pipeline.add_always(std::make_unique<TypedEventPipelineStage<TestEvent>>());
+
+	REQUIRE(pipeline.always_count() == 1);
+}
+
+TEST_CASE("register multiple stages", "[events]")
+{
+	TestEventPipeline pipeline{};
+
+	REQUIRE(pipeline.always_count() == 0);
+	REQUIRE(pipeline.once_count() == 0);
+
+	pipeline
+	    .add_once(std::make_unique<TypedEventPipelineStage<TestEvent>>())
+	    .add_once(std::make_unique<TypedEventPipelineStage<TestEvent>>())
+	    .add_always(std::make_unique<TypedEventPipelineStage<TestEvent>>())
+	    .add_always(std::make_unique<TypedEventPipelineStage<TestEvent>>())
+	    .add_always(std::make_unique<TypedEventPipelineStage<TestEvent>>());
+
+	REQUIRE(pipeline.always_count() == 3);
+	REQUIRE(pipeline.once_count() == 2);
+	REQUIRE(pipeline.total_stage_count() == 5);
+}
+
+TEST_CASE("stages are executed in the correct order", "[events]")
+{
+	TestEventPipeline pipeline{};
+
+	uint32_t execution_index{0};
+
+	pipeline
+	    .add_once(std::make_unique<TestStage>(execution_index, 0))
+	    .add_once(std::make_unique<TestStage>(execution_index, 1))
+	    .add_always(std::make_unique<TestStage>(execution_index, 2))
+	    .add_always(std::make_unique<TestStage>(execution_index, 3))
+	    .add_always(std::make_unique<TestStage>(execution_index, 4));
+
+	// assertions inside TestStage
+	pipeline.process();
+
+	// on a second pass we expect only the "then" stages to be processed
+	execution_index = 2;
+
+	pipeline.process();
+}
+
+TEST_CASE("event observers are executed in the correct order", "[events]")
+{
+	struct EventOne
+	{
+		uint32_t value{4};
+	};
+
+	struct EventTwo
+	{
+		uint32_t value{1};
+	};
+
+	struct EventThree
+	{
+		uint32_t value{56};
+	};
+
+	TestEventPipeline pipeline{};
+
+	pipeline
+	    .add_once(std::make_unique<TypedEventPipelineStage<EventOne>>())
+	    .add_always(std::make_unique<TypedEventPipelineStage<EventTwo>>())
+	    .add_always(std::make_unique<TypedEventPipelineStage<EventThree>>());
+
+	uint32_t one_execution_count{0};
+	uint32_t two_execution_count{0};
+	uint32_t three_execution_count{0};
+
+	pipeline
+	    .each<EventOne>([&](const EventOne &event) {
+		    REQUIRE(event.value == 4);
+		    one_execution_count++;
+	    })
+	    .each<EventTwo>([&](const EventTwo &event) {
+		    REQUIRE(event.value == 1);
+		    two_execution_count++;
+	    })
+	    .each<EventThree>([&](const EventThree &event) {
+		    REQUIRE(event.value == 56);
+		    three_execution_count++;
+	    });
+
+	pipeline.process();
+	pipeline.process();
+	pipeline.process();
+	pipeline.process();
+
+	REQUIRE(one_execution_count == 1);
+	REQUIRE(two_execution_count == 4);
+	REQUIRE(three_execution_count == 4);
```

Comment:
I see, the number of calls of each event is checked, but how is the correct order verified?